carbon-design-system/carbon-addons-iot-react
# feat(datetimepicker): close dropdown on click outside

Closes #3642 

**Summary**

- Add support to close dropdown when a user clicks outside

**Change List (commits, features, bugs, etc)**

- Add mouse and keyboard events listener to the `DateTimePicker(V2)` components
- Add tests
- Update snapshots (`DashboardGrid` e2e snapshot had 2 cards with open date time picker, due to new changes only 1 of them is displayed now)

**Acceptance Test (how to verify the PR)**

- _Saved value_
- Go to [Date time range selector / Default](https://deploy-preview-3671--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-date-time-range-selector--default)
- Select some date range
- Click on Apply button to save changes
- Select another date range
- Without saving, click outside to close dropdown
- Verify that last saved value was restored

</br>

- _Selected preset_ 
- Go to [Date time range selector / Selected preset](https://deploy-preview-3671--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-date-time-range-selector--selected-preset)
- Select some value
- Click outside of the component or click on input field
- Verify that dropdown had been closed and default preset was restored

</br>

- _Default relative value_
- Go to [Date time range selector / Selected relative](https://deploy-preview-3671--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-date-time-range-selector--selected-relative)
- Select some value
- Click outside of the component or click on input field
- Verify that dropdown had been closed and default relative value was restored

</br>

- _DateTimePickerV2_
- Go to [DateTimePickerV2 / Selected absolute](https://deploy-preview-3671--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--selected-absolute)
- Pick some dates from calendar
- Click outside of the component or click on input field
- Verify that dropdown had been closed and default absolute value was restored

</br>

- _Single select_
- Go to [DateTimePickerV2 / Single select with new time spinner](https://deploy-preview-3671--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--single-select)
- Select some date from calendar
- Save value by clicking Apply button
- Select some other date
- Without saving changes, click outside of the component
- Verify that dropdown had been close and last saved value was restored

**Regression Test (how to make sure this PR doesn't break old functionality)**

- tests here



**Things to look for during review**
